<launch>

  <!-- Link to the custom libsub (LD_LIBRARY_PATH is searched before ld.so.conf.d, so the custom libusb will be used instead of the system default-->
  <env name="LD_LIBRARY_PATH" value="/opt/mvIMPACT_acquire_libusb:$(env LD_LIBRARY_PATH)" />

  <!-- Node delay for multiple cameras -->
  <arg name="node_start_delay" default="0"/>

  <!-- Common Interface -->
  <arg name="device"/>
  <arg name="rate" default="30"/>

  <!-- Camera Settings --> //described in config file
  <arg name="serial" default="$(arg device)"/>
  <arg name="camera_name" default="mv_$(arg serial)"/>
  <arg name="camera" default="$(arg camera_name)"/>
  <arg name="frame_id" default="$(arg camera)"/>
  <arg name="calib_url" default="file://${ROS_HOME}/camera_info/$(arg camera_name).yaml"/>
  <arg name="fps" default="$(arg rate)"/>
  <arg name="color" default="false"/>
  <arg name="aec" default="true"/> //true for autoexposure
  <arg name="cbm" default="false"/> //binnig
  <arg name="ctm" default="3"/> //trigger mode - if 3 - 4 using expose values, otherwise grows lineary upon expose time
  <arg name="cts" default="0"/> // trigger source (-1 is also fine)
  <arg name="dcfm" default="0"/> //dark current filter - case of high darkness
  <arg name="hdr" default="false"/> // when using hdr change gain_db to 0, otherwise effect will be reversed
  <arg name="wbp" default="-1"/> //white balance
  <arg name="expose_us" default="5000"/>
  <arg name="gain_db" default="12.0"/> //max is 32 - ISO from expose triangle, but fro us limited to 12
  <arg name="boost" default="false"/>
  <arg name="mm" default="0"/> // mirror mode
  <arg name="expose_upper_limit_us" default="60000"/> //lower exp level to limit the camera
  <arg name="des_grey_value" default="128"/>

  <!-- Arguments for theora quality, images being compressed into "stream" -->
  <arg name="compressed_jpeg_quality" default="90" />
  <arg name="theora_keyframe_frequency" default="60"  />
  <arg name="theora_target_bitrate" default="50000"  />
  <arg name="theora_quality" default="8"  />
  <arg name="theora_optimize_for" default="0"  />

  <!-- Node Settings -->
  <arg name="plugin" default="false"/>
  <arg name="manager" default="bluefox_manager"/>
  <arg name="output" default="screen"/>
  <arg name="proc" default="false"/>
  <arg name="view" default="false"/> // true for automatic view window load if false call rosrun rqt_image_view rqt_image_view(options awailable)
  <arg name="calib" default="false"/>

  <!-- Standalone -->
  <arg unless="$(arg plugin)"
    name="option" value="standalone bluefox2/SingleNodelet"/>

  <!-- Manager -->
  <arg if="$(arg plugin)"
    name="option" value="load bluefox2/SingleNodelet $(arg manager)"/>

  <!-- Nodelet -->
  <node pkg="nodelet" type="nodelet" name="$(arg camera)"
    args="$(arg option)" output="$(arg output)" respawn="true" launch-prefix="bash -c 'sleep $(arg node_start_delay); $0 $@' ">
    <param name="identifier" type="string" value="$(arg serial)"/>
    <param name="frame_id" type="string" value="$(arg frame_id)"/>
    <param name="camera_name" type="string" value="$(arg camera_name)"/>
    <param name="calib_url" type="string" value="$(arg calib_url)"/>
    <param name="fps" type="double" value="$(arg fps)"/>
    <param name="color" type="bool" value="$(arg color)"/>
    <param name="aec" type="bool" value="$(arg aec)"/>
    <param name="cbm" type="bool" value="$(arg cbm)"/>
    <param name="ctm" type="int" value="$(arg ctm)"/>
    <param name="cts" type="int" value="$(arg cts)"/>
    <param name="dcfm" type="int" value="$(arg dcfm)"/>
    <param name="hdr" type="bool" value="$(arg hdr)"/>
    <param name="wbp" type="int" value="$(arg wbp)"/>
    <param name="expose_us" type="int" value="$(arg expose_us)"/>
    <param name="gain_db" type="double" value="$(arg gain_db)"/>
    <param name="boost" type="bool" value="$(arg boost)"/>
    <param name="mm" type="int" value="$(arg mm)"/>
    <param name="expose_upper_limit_us" type="int" value="$(arg expose_upper_limit_us)"/>
    <param name="des_grey_value" type="int" value="$(arg des_grey_value)"/>

    <!-- Params for theora quality, images being compressed into "stream" -->
    <param name="image_raw/compressed/jpeg_quality" type="int" value="$(arg compressed_jpeg_quality)" />
    <param name="image_raw/theora/keyframe_frequency" value="$(arg theora_keyframe_frequency)"  />
    <param name="image_raw/theora/target_bitrate" value="$(arg theora_target_bitrate)"  />
    <param name="image_raw/theora/quality" value="$(arg theora_quality)"  />
    <param name="image_raw/theora/optimize_for" value="$(arg theora_optimize_for)"  />
  </node>

  <!--Proc-->
  <arg if="$(arg plugin)"
    name="proc_option" value="load image_proc/rectify $(arg manager)"/>
  <arg unless="$(arg plugin)"
    name="proc_option" value="standalone image_proc/rectify"/>
  <node if="$(arg proc)"
    pkg="nodelet" type="nodelet" name="rectify_mono" ns="$(arg camera)"
    args="$(arg proc_option)">
    <remap from="image_mono" to="image_raw"/>
    <remap from="image_rect" to="image_rect_mono"/>
  </node>

  <!-- View -->
  <arg name="image" default="image_raw"/>
  <node if="$(arg view)"
    pkg="image_view" type="image_view" name="viewer"
    output="$(arg output)" ns="$(arg camera)" args="image:=$(arg image)">
  </node>
</launch>
